src/index.ts

```diff
@@ -93,13 +93,14 @@ const NanoClamp = ({
     updateTextRefs(textPlusEllipsis)
   }, [ellipsis, hasText, lines, text])
 
-  useEffect(() => {
+  useLayoutEffect(() => {
     clampLines()
 
-    const clampLinesDebounced = debounceFn(clampLines, debounce)
-    window.addEventListener('resize', clampLinesDebounced)
-
-    return () => window.removeEventListener('resize', clampLinesDebounced)
+    if (elementRef.current != null) {
+      const observer = new ResizeObserver(debounceFn(clampLines, debounce))
+      observer.observe(elementRef.current)
+      return () => observer.disconnect()
+    }
```

Comment:
**Thought**: maybe we could move `observer` to a `let` outside of this if statement, and move the return statement outside of the if? And then do something like `observe?.disconnect()`.

So we're sure we always disconnect even if the component happens to unmount. Shouldn't, but just to be on the safe side.

Reply:
I'm not sure if I understand your concern. Either the `observer` is created, then a cleanup function is returned that disconnects it when unmounted. Or this block is not executed (there is no observer), then no cleanup function is returned. So we are on the safe side. I guess it's more a question of readability.

Reply:
I've rewritten the code to use an early return. I think that way it is more clear what is happening.
(By the way, I believe we could move that early return even before the `clampLines()` call in line 97. However, I'm not sure if I miss something ...)